save-preprocessor/src/main/kotlin/org/cqfn/save/preprocessor/controllers/DownloadProject.kt

```diff
@@ -36,42 +37,46 @@ class DownloadProject {
-        val userCredentials = if (gitRepository.username != null && gitRepository.password != null) {
-            UsernamePasswordCredentialsProvider(gitRepository.username, gitRepository.password)
-        } else {
-            CredentialsProvider.getDefault()
-        }
-        try {
-            Git.cloneRepository()
-                .setURI(gitRepository.url)
-                .setCredentialsProvider(userCredentials)
-                .setDirectory(tmpDir)
-                .call().use {
-                    log.info("Repository cloned: ${gitRepository.url}")
-                    // TODO post request to orchestrator
-                    return Mono.just(ResponseEntity("Cloned", HttpStatus.ACCEPTED))
-                }
-        } catch (exception: Exception) {
-            return when (exception) {
-                is InvalidRemoteException,
-                is TransportException,
-                is GitAPIException -> {
-                    log.warn("Error with git API while cloning ${gitRepository.url} repository", exception)
-                    Mono.just(ResponseEntity(exception.stackTraceToString(), HttpStatus.BAD_REQUEST))
-                }
-                else -> {
-                    log.warn("Cloning ${gitRepository.url} repository failed", exception)
-                    Mono.just(ResponseEntity(exception.stackTraceToString(), HttpStatus.INTERNAL_SERVER_ERROR))
+        return Mono.just(ResponseEntity("Cloned", HttpStatus.ACCEPTED))
+            .subscribeOn(Schedulers.boundedElastic())
+            .also {
+                it.subscribe {
+                    val urlHash = gitRepository.url.hashCode()
+                    val tmpDir = File("$volumes/$urlHash")
+                    if (tmpDir.exists()) {
+                        tmpDir.deleteRecursively()
+                        log.info("For ${gitRepository.url} repository: dir $urlHash was deleted")
+                    }
+                    tmpDir.mkdirs()
+                    log.info("For ${gitRepository.url} repository: dir $urlHash was created")
+                    val userCredentials = if (gitRepository.username != null && gitRepository.password != null) {
+                        UsernamePasswordCredentialsProvider(gitRepository.username, gitRepository.password)
+                    } else {
+                        CredentialsProvider.getDefault()
+                    }
+                    try {
+                        Git.cloneRepository()
+                            .setURI(gitRepository.url)
+                            .setCredentialsProvider(userCredentials)
+                            .setDirectory(tmpDir)
+                            .call().use {
+                                log.info("Repository cloned: ${gitRepository.url}")
+                                // TODO post request to orchestrator
+                            }
+                    } catch (exception: Exception) {
+                        tmpDir.deleteRecursively()
```

Comment:
Should be in `finally`

Reply:
Why it should be in finally? We delete directory only if there was an exception.